internal/app/go-filecoin/internal/submodule/storage_protocol_submodule.go

```diff
@@ -2,22 +2,105 @@ package submodule
+	gsnet "github.com/ipfs/go-graphsync/network"
+	gsstoreutil "github.com/ipfs/go-graphsync/storeutil"
+	"github.com/ipfs/go-hamt-ipld"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/pkg/errors"
+
+	"github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/paths"
+	"github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/plumbing/msg"
+	storagemarketconnector "github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/storage_market_connector"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/piecemanager"
+	fcaddr "github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/wallet"
 )
 
-// StorageProtocolSubmodule enhances the `Node` with "Storage" protocol capabilities.
+// StorageProtocolSubmodule enhances the node with storage protocol
+// capabilities.
 type StorageProtocolSubmodule struct {
-	StorageAPI *storage.API
-
-	// Storage Market Interfaces
-	StorageMiner *storage.Miner
+	StorageClient   iface.StorageClient
+	StorageProvider iface.StorageProvider
 }
 
 // NewStorageProtocolSubmodule creates a new storage protocol submodule.
-func NewStorageProtocolSubmodule(ctx context.Context) (StorageProtocolSubmodule, error) {
-	return StorageProtocolSubmodule{
-		// StorageAPI: nil,
-		// StorageMiner: nil,
+func NewStorageProtocolSubmodule(
+	ctx context.Context,
+	minerAddr fcaddr.Address,
+	clientAddr fcaddr.Address,
+	c *ChainSubmodule,
+	m *MessagingSubmodule,
+	mw *msg.Waiter,
+	pm piecemanager.PieceManager,
+	wlt *wallet.Wallet,
+	h host.Host,
+	ds datastore.Batching,
+	bs blockstore.Blockstore,
+	repoPath string,
+	wg storagemarketconnector.WorkerGetter) (*StorageProtocolSubmodule, error) {
+
+	ma, err := address.NewFromBytes(minerAddr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	ca, err := address.NewFromBytes(clientAddr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	pnode := storagemarketconnector.NewStorageProviderNodeConnector(ma, c.State, m.Outbox, mw, pm, wg, wlt)
+	cnode := storagemarketconnector.NewStorageClientNodeConnector(hamt.CSTFromBstore(bs), c.State, mw, wlt, m.Outbox, ca, wg)
+
+	pieceStagingPath, err := paths.PieceStagingDir(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure pieces directory exists
+	err = os.MkdirAll(pieceStagingPath, 0700)
+	if err != nil {
+		return nil, err
+	}
+
+	fs, err := filestore.NewLocalFileStore(filestore.OsPath(pieceStagingPath))
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure miner-address is in datastore because graphsyncimpl requires it
+	err = ds.Put(datastore.NewKey("miner-address"), minerAddr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	graphsyncNetwork := gsnet.NewFromLibp2pHost(h)
+	bridge := ipldbridge.NewIPLDBridge()
+	loader := gsstoreutil.LoaderForBlockstore(bs)
+	storer := gsstoreutil.StorerForBlockstore(bs)
+	gsync := graphsync.New(ctx, graphsyncNetwork, bridge, loader, storer)
```

Comment:
I'm suspicious we are going to get into a lot of trouble starting two graphsync protocol instances on the same host and datastore.  Even if we don't do you think it would be better practice for us to use the same graphsync protocol?  Is this separation intended to prevent interference between chain sync and piece transfer?

Reply:
This was just added here because it it inaccessible from the Syncer. That's not really a good excuse. I'll work on breaking it out so it can be shared.

Reply:
I've moved graphsync initialization into the network submodule so the syncer and storage protocol can share it.

Reply:
Graphsync construction has been moved to the network submodule and the graphsync exchange is now shared between the syncer and the storage protocol.